src/module/SynchrotronRadiation.cpp

```diff
@@ -129,25 +147,58 @@ void SynchrotronRadiation::process(Candidate *candidate) const {
 		return;
 
 	// draw photons up to the total energy loss
+	// if maximumSamples is reached before that, compensate the total energy afterwards
 	Random &random = Random::instance();
+	double dE0 = dE;
+	std::vector<double> energies;
+	int counter = 0;
 	while (dE > 0) {
 		// draw random value between 0 and maximum of corresponding cdf
 		// choose bin of s where cdf(x) = cdf_rand -> x_rand
 		size_t i = random.randBin(tabCDF); // draw random bin (upper bin boundary returned)
 		double binWidth = (tabx[i] - tabx[i-1]);
 		double x = tabx[i-1] + random.rand() * binWidth; // draw random x uniformly distributed in bin
-		double Egamma = x * Ecrit;
+		double Ephoton = x * Ecrit;
 
 		// if the remaining energy is not sufficient check for random accepting
-		if (Egamma > dE)
-			if (random.rand() > (dE / Egamma))
+		if (Ephoton > dE) {
+			if (random.rand() > (dE / Ephoton))
 				break; // not accepted
+		}
+
+		// only activate the "per-step" sampling if maximumSamples is explicitly set.
+		if (maximumSamples > 0) {
+			if (counter >= maximumSamples) 
+				break;			
+		}
 
-		// create synchrotron photon and repeat with remaining energy
-		dE -= Egamma;
-		Vector3d pos = random.randomInterpolatedPosition(candidate->previous.getPosition(), candidate->current.getPosition());
-		if (Egamma > secondaryThreshold) // create only photons with energies above threshold
-			candidate->addSecondary(22, Egamma, pos);
+		// store energies in array
+		energies.push_back(Ephoton);
+
+		// energy loss
+		dE -= Ephoton;
+
+		// counter for sampling break condition;
+		counter++;
+	}
+
+	// while loop before gave total energy which is just a fraction of the required
+	double w1 = 1;
+	if (maximumSamples > 0 && dE > 0)
+		w1 = 1. / (1. - dE / dE0); 
+
+	// loop over sampled photons and attribute weights accordingly
+	for (int i = 0; i < energies.size(); i++) {
+		double Ephoton = energies[i];
+		double f = Ephoton / (E - dE0);
+		double w = w0 * w1 / pow(f, thinning);
```

Comment:
The last part of the weight `1/pow(f, thinning)` then accounts for the fact that only a fraction of the sampled photons is accepted, as in the modules with thinning, right?

Reply:
Yes. This is essentially the usual thinning method used in the other EM modules. It doesn't help a lot in most cases, but combined with the sampling, it increases the performance significantly.